arangod/IResearch/IResearchAqlAnalyzer.cpp

```diff
@@ -508,13 +509,91 @@ namespace iresearch {
   return make_slice(builder->slice());
 }
 
+// return the calculation node
+ExecutionNode* tryOptimize(ExecutionNode* node) {
+  if (node == nullptr || node->getType() != ExecutionNode::RETURN) {
+    return nullptr;
+  }
+
+  auto& deps = node->getDependencies();
+  if (deps.size() == 1 && deps[0]->getType() == ExecutionNode::NodeType::CALCULATION) {
+    auto calcNode = deps[0];
+
+    auto& deps2 = calcNode->getDependencies();
+    if (deps2.size() == 1 && deps2[0]->getType() == ExecutionNode::NodeType::SINGLETON) {
+      return calcNode;
+    }
+  }
+
+  return nullptr;
+}
+
+AqlAnalyzer::QueryOptimizer::QueryOptimizer(AqlAnalyzer* analyzer) : _analyzer(analyzer) {
+  _result = _analyzer->_itemBlockManager.requestBlock(1,1);
+}
+
+bool AqlAnalyzer::QueryOptimizer::optimize(irs::string_ref const& field) {
+  if (!_nodeToOptimize) {
+    ExecutionNode* execNode = tryOptimize(_analyzer->_plan->root());
+    _nodeToOptimize = dynamic_cast<CalculationNode*>(execNode);
```

Comment:
dynamic_cast only in debug, add TRI_ASSERT that execNode is NODE_TYPE_CALCULATION